### PR TITLE
mgr/volumes: periodically check for async work

### PR DESCRIPTION
**Pull request by vshankar**
Useful when "manually" assigning work - with this fix there is not need to bounce ceph-mgr to kick async threads.

Fixes: http://tracker.ceph.com/issues/61867

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by vshankar** [comment]
GChat update: looks like the race window is easily hit with this change - https://pulpito.ceph.com/vshankar-2023-08-17_09:35:41-fs:volumes-wip-vshankar-testing-20230816.105627-testing-default-smithi/

@kotreshhr I don't have any bright idea apart from making this configurable and then disabling it from the test yaml. Thoughts?

**Comment by kotreshhr** [comment]
> GChat update: looks like the race window is easily hit with this change - https://pulpito.ceph.com/vshankar-2023-08-17_09:35:41-fs:volumes-wip-vshankar-testing-20230816.105627-testing-default-smithi/
> 
> @kotreshhr I don't have any bright idea apart from making this configurable and then disabling it from the test yaml. Thoughts?

In the teuthology tests, it's easily hit because the fs is destroyed using `ceph fs fail` and not `ceph  fs volume rm` command. I think reason is that the `ceph fs fail` is faster.  The `fs fail` immediately makes mds non-joinable and fails all the mdses.  The `fs volume rm` does cancel the threads before removing the volume.
To fix the test case, instead of disabling the test for this, we can disable/enable the mgr/volumes plugin in tearDown for CephFS tests? But this might slowdown the overall test running time?

My only concern is that  this is not configurable and what if we start seeing the users hitting this race window. I am not very sure about the user base of multi-fs setups or how frequently they destroy file-systems using `ceph fs fail` ?

Or instead of this fix, can we just use the workaround of disabling and enabling mgr/volumes plugin module for purge thread to pick up the new manual entries?

**Comment by vshankar** [comment]
> > GChat update: looks like the race window is easily hit with this change - https://pulpito.ceph.com/vshankar-2023-08-17_09:35:41-fs:volumes-wip-vshankar-testing-20230816.105627-testing-default-smithi/
> > @kotreshhr I don't have any bright idea apart from making this configurable and then disabling it from the test yaml. Thoughts?
> 
> In the teuthology tests, it's easily hit because the fs is destroyed using `ceph fs fail` and not `ceph fs volume rm` command. I think reason is that the `ceph fs fail` is faster. The `fs fail` immediately makes mds non-joinable and fails all the mdses. The `fs volume rm` does cancel the threads before removing the volume. To fix the test case, instead of disabling the test for this, we can disable/enable the mgr/volumes plugin in tearDown for CephFS tests? But this might slowdown the overall test running time?

mgr/volumes is an "always-on" module and can't be disabled.

> 
> My only concern is that this is not configurable and what if we start seeing the users hitting this race window. I am not very sure about the user base of multi-fs setups or how frequently they destroy file-systems using `ceph fs fail` ?

That's a fair point, and the underlying issues needs to be fixed, I.e., client should receive back some errno for the file system call when the underlying file system (and the MDSs) are removed. I'll create a tracker for that.

> 
> Or instead of this fix, can we just use the workaround of disabling and enabling mgr/volumes plugin module for purge thread to pick up the new manual entries?

Would have been ideal but the module cannot be disabled.

**Comment by vshankar** [comment]
I pushed an update - @kotreshhr PTAL. Haven't tested the test case for now, but the changes work locally.

**Review by kotreshhr — APPROVED** [review]
Looks good to me.

File: src/pybind/mgr/volumes/fs/async_job.py [hunk 1]
```diff
@@ -105,6 +104,9 @@ class AsyncJobs(threading.Thread):
     via `should_cancel()` lambda passed to `execute_job()`.
     """
 
+    # not made configurable on purpose
+    WAKEUP_TIMEOUT = 5.0
```
Comment (dparmar18) [inline src/pybind/mgr/volumes/fs/async_job.py:108]: just curious to know why we didn't go with the option to make it configurable?

Comment (vshankar) [reply src/pybind/mgr/volumes/fs/async_job.py:108]: since it hardly matters if a "manual" work is picked up now or 5s later.

**Comment by vshankar** [comment]
Fixed a bug where the async job wakes up but finds `self.q` empty since when manually placing a work. Fix is not populate `self.q` with the list of volumes.

**Comment by kotreshhr** [comment]
> Fixed a bug where the async job wakes up but finds `self.q` empty since when manually placing a work. Fix is not populate `self.q` with the list of volumes.

Ah! Good catch! I missed that part.  Did the test added catch this bug?

File: src/pybind/mgr/volumes/fs/async_job.py [hunk 2]
```diff
@@ -41,10 +42,17 @@ def run(self):
                             log.info("thread [{0}] terminating due to reconfigure".format(thread_name))
                             self.async_job.threads.remove(self)
                             return
+                        timo = self.async_job.wakeup_timeout
```
Comment (dparmar18) [inline src/pybind/mgr/volumes/fs/async_job.py:45]: nit: s/timo/timeout

Comment (vshankar) [reply src/pybind/mgr/volumes/fs/async_job.py:45]: I've always used timo when I first ran into a namespace collision with timoeout - this was way back ;P

**Comment by vshankar** [comment]
> Ah! Good catch! I missed that part. Did the test added catch this bug?

yes.

**Comment by mchangir** [comment]
http://pulpito.front.sepia.ceph.com/mchangir-2023-09-12_05:40:22-fs-wip-mchangir-testing-20230908.140927-testing-default-smithi/

**Pull request merged**